### ISSUE TITLE
V3 longer than V2 (MAC OS)

### ISSUE THREAD

**Issue by Afraz496** [HPCE/hpce-2017-cw3#68]
Platform: MAC OS.

So after finishing v3, I found the time taken for n = 1000 (where OpenCL should be faster than the software version) is in fact longer. I imagine it is the same for v4 as I haven't tested it but it showed up as an error on my logs. 

1) time (cat /tmp/world.bin | bin/step_world 0.1 1  1 > /dev/null) 
Loaded world with w=1000, h=1000
Stepping by dt=0.1 for n=1

real	0m0.466s
user	0m0.443s
sys	0m0.023s

2) time (cat /tmp/world.bin | bin/ak8314/step_world_v3_opencl 0.1 1  1 > /dev/null)
Loaded world with w=1000, h=1000
Stepping by dt=0.1 for n=1
Found 1 platforms
  Platform 0 : Apple
Choosing platform 0
Found 2 devices
  Device 0 : Intel(R) Core(TM) i5-3230M CPU @ 2.60GHz
  Device 1 : HD Graphics 4000
Choosing device 1

real	0m0.524s
user	0m0.435s
sys	0m0.038s

As you can see the real-time for software is shorter than that for v3. v1 and v2 take (around the same) less time than the software for n = 1000:

V1
Loaded world with w=1000, h=1000
Stepping by dt=0.1 for n=1

real	0m0.461s
user	0m0.437s
sys	0m0.020s

V2
Loaded world with w=1000, h=1000
Stepping by dt=0.1 for n=1

real	0m0.467s
user	0m0.442s
sys	0m0.021s

I initially thought this was a platform issue but my logs didn't pass the time step. 

EDIT: for V4

Even ran the result for v4, got an abnormally large time value:

Loaded world with w=1000, h=1000
Stepping by dt=0.1 for n=1
Found 1 platforms
  Platform 0 : Apple
Choosing platform 0
Found 2 devices
  Device 0 : Intel(R) Core(TM) i5-3230M CPU @ 2.60GHz
  Device 1 : HD Graphics 4000
Choosing device 1

real	0m0.772s
user	0m0.440s
sys	0m0.040s

**Comment by Afraz496** [comment]
Further Edit, so running the tests on n = 5000 gives me the desired result:

step_world:

real	3m36.976s
user	3m35.888s
sys	0m1.036s

v3:

real	1m18.157s
user	0m11.083s
sys	0m20.031s

v4

real	0m43.585s
user	0m10.884s
sys	0m0.738s

That would mean my implementation is correct then?

**Comment by m8pple** [comment]
You're probably thinking of it in slightly the wrong way - performance is usually
not a correctness issue, as it is a non-functional requirement (*ahem* - last Friday's lecture).

So where you say "where OpenCL should be faster than the software version", is that actually true?
What hardware is being used, and how does it compare to the software (in raw performance terms)?
How efficient is the Apple provider compared to (say) the NVidia AWS provider?

It's noticeable in the first set of results that the times are quite small, less than a second. This
needs to incorporate the process startup cost, OpenCL initialisation cost, and potentially the
time taken to compile the OpenCL kernel.

I'd also query the assertion in the first set of results where you say:

> So after finishing v3, I found the time taken for n = 1000

versus the output:
```
Stepping by dt=0.1 for n=1
```

To answer these questions I'd be inclined to look at scaling results. You can assume that there
is a fixed-cost for starting the process and initialising OpenCL, and a variable (linear) cost for
each time-step:
- What experiments or graphs (with only 4 or 5 data-points) could you use to explore this?
- How do you test your assumptions and expectations?

**Comment by Afraz496** [comment]
Many thanks!! I realized I typed up n = 1000 at the end, my apologies. So for testing at n = 1000 my results are now: 

V3
time (cat /tmp/world.bin | bin/ak8314/step_world_v3_opencl 0.1 1000  1 > /dev/null)
Loaded world with w=1000, h=1000
Stepping by dt=0.1 for n=1000
Found 1 platforms
  Platform 0 : Apple
Choosing platform 0
Found 2 devices
  Device 0 : Intel(R) Core(TM) i5-3230M CPU @ 2.60GHz
  Device 1 : HD Graphics 4000
Choosing device 1

real	0m3.834s
user	0m0.634s
sys	0m0.789s

V4
time (cat /tmp/world.bin | bin/ak8314/step_world_v4_double_buffered 0.1 1000  1 > /dev/null)
Loaded world with w=1000, h=1000
Stepping by dt=0.1 for n=1000
Found 1 platforms
  Platform 0 : Apple
Choosing platform 0
Found 2 devices
  Device 0 : Intel(R) Core(TM) i5-3230M CPU @ 2.60GHz
  Device 1 : HD Graphics 4000
Choosing device 1

real	0m1.806s
user	0m0.487s
sys	0m0.099s

V5
time (cat /tmp/world.bin | bin/ak8314/step_world_v5_packed_properties 0.1 1000  1 > /dev/null)
Loaded world with w=1000, h=1000
Stepping by dt=0.1 for n=1000
Found 1 platforms
  Platform 0 : Apple
Choosing platform 0
Found 2 devices
  Device 0 : Intel(R) Core(TM) i5-3230M CPU @ 2.60GHz
  Device 1 : HD Graphics 4000
Choosing device 1

real	0m1.391s
user	0m0.496s
sys	0m0.114s

which is what I expected. I was running this on a Macbook Pro 13" Retina (Mid 2013), it has Intel HD Graphics 4000 1536 MB.
